File: csharp/ql/src/Security Features/CWE-730/RegexInjection.cs
using System;
using System.Web;
using System.Text.RegularExpressions;

public class RegexInjectionHandler : IHttpHandler
{

    public void ProcessRequest(HttpContext ctx)
    {
        string name = ctx.Request.QueryString["name"];
        string userInput = ctx.Request.QueryString["userInput"];

        // BAD: Unsanitized user input is used to construct a regular expression
        new Regex("^" + name + "=.*$").Match(userInput);

        // GOOD: User input is sanitized before constructing the regex
        string safeName = Regex.Escape(name);
        new Regex("^" + safeName + "=.*$").Match(userInput);
    }
}

File: java/ql/src/experimental/Security/CWE/CWE-730/RegexInjection.java
package com.example.demo;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class DemoApplication {

    @GetMapping("/string1")
    public String string1(@RequestParam(value = "input", defaultValue = "test") String input,
                          @RequestParam(value = "pattern", defaultValue = ".*") String pattern) {
        // BAD: Unsanitized user input is used to construct a regular expression
        if (input.matches("^" + pattern + "=.*$"))
            return "match!";

        return "doesn't match!";
    }

    @GetMapping("/string2")
    public String string2(@RequestParam(value = "input", defaultValue = "test") String input,
                          @RequestParam(value = "pattern", defaultValue = ".*") String pattern) {
        // GOOD: User input is sanitized before constructing the regex
        if (input.matches("^" + escapeSpecialRegexChars(pattern) + "=.*$"))
            return "match!";

        return "doesn't match!";
    }

    Pattern SPECIAL_REGEX_CHARS = Pattern.compile("[{}()\\[\\]><-=!.+*?^$\\\\|]");

    String escapeSpecialRegexChars(String str) {
        return SPECIAL_REGEX_CHARS.matcher(str).replaceAll("\\\\$0");
    }
}